---
title: "locuszoomr"
author: "Myles Lewis"
output:
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true
vignette: >
  %\VignetteIndexEntry{locuszoomr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE
)
library(locuszoomr)
```

# Introduction

The `locuszoomr` package allows users to produce publication ready gene locus 
plots very similar to those produced by the web interface locuszoom 
(http://locuszoom.org), but running purely locally in R. It provides 
customisation to the plots.

These gene annotation plots are produced via R base graphics system. For users
who prefer the grid system we recommend looking at Bioconductor packages `Gviz`
or `ggbio`.

# Installation

Bioconductor packages `ensembldb` and an Ensembl database package are required 
before installation.

```{r eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("ensembldb")
BiocManager::install("EnsDb.Hsapiens.v75")
```

Install from Github
```{r eval = FALSE}
devtools::install_github("myles-lewis/locuszoomr")
library(locuszoomr)
library(EnsDb.Hsapiens.v75)
```

`locuszoomr` automatically leverages the `LDlinkR` package to query 1000 Genomes 
for linkage disequilibrium (LD) across SNPs. In order to make use of this API
function you will need a personal access token (see the `LDlinkR` vignette), 
available from the LDlink website https://ldlink.nci.nih.gov/?tab=apiaccess.

Requests to LDlink are cached using the `memoise` package, to reduce API 
requests. This is helpful when modifying plots for aesthetic reasons.

# Example locus plot

The quick example below uses a small subset of a GWAS dataset incorporated into
the package as a demo. The dataset is from a genetic study on Systemic Lupus
Erythematosus (SLE) by Bentham et al (2015). The data format is shown below.

```{r}
data(SLE_gwas_sub)  ## limited subset of data from SLE GWAS
head(SLE_gwas_sub)
```

We plot a locus from this dataset by extracting a subset of the data using the
`locus()` function.

```{r warning=FALSE, message=FALSE, fig.dim=c(8, 7)}
library(EnsDb.Hsapiens.v75)
loc <- locus(SLE_gwas_sub, gene = 'UBE2L3', flank = 1e5)
summary(loc)
plot(loc)
```

# Obtaining LD information

Once an API personal access token has been obtained, the LDlink API can be
called using the function `link_LD()` to retrieve LD (linkage disequilibium)
information at the locus which is overlaid on the locus plot. This is shown as a
colour overlay showing the level of $r^2$ between SNPs and the index SNP which
defaults to the SNP with the lowest p-value (or the SNP can be specified
manually).

```{r eval = FALSE}
# Locus plot using SLE GWAS data from Bentham et al 2015
# FTP download full summary statistics from
# https://www.ebi.ac.uk/gwas/studies/GCST003156
library(data.table)
SLE_gwas <- fread('../bentham_2015_26502338_sle_efo0002690_1_gwas.sumstats.tsv')
loc <- locus(SLE_gwas, gene = 'UBE2L3', flank = 1e5)
loc <- link_LD(loc, LDtoken = "your_token")
plot(loc)
```

The subset of GWAS data included in the `locuszoomr` package has LD data already
acquired from LDlink which is included in the `r2` column. This can be plotted
by setting `LD = "r2"`. This method allows users to add their own LD information
from their own datasets to loci.

```{r fig.dim=c(8, 7)}
loc <- locus(SLE_gwas_sub, gene = 'UBE2L3', flank = 1e5, LD = "r2")
plot(loc)
```

# Plot customisation

Various plotting options can be customised through arguments via the call to
`plot()`. Plot borders can be set using `border = TRUE`. The chromosome position
$x$ axis labels can be placed under the top or bottom plots using 
`xtick = "top"` or `"bottom"`.

See the help page at `?plot.locus` for more details.

# Customising gene tracks

The gene tracks can be also customised with colours and gene label text
position. See the help page at `?plot.locus` for more details.

```{r eval = FALSE}
# Filter by gene biotype
plot(loc, filter_gene_biotype = "protein_coding")

# Custom selection of genes using gene names
plot(loc, filter_gene_name = c('UBE2L3', 'RIMBP3C', 'YDJC', 'PPIL2',
                                     'PI4KAP2', 'MIR301B'))
```

# Arrange multiple plots

`locuszoomr` uses `graphics::layout` to arrange plots. To layout multiple locus
plots side by side, use the function `multi_layout()` to set the number of locus
plots per row and column. Then set the argument `use_layout = FALSE` when
plotting locus objects. `multi_layout()` returns a list of original `par()`
parameters so they can be reset after plotting has finished.

```{r eval = FALSE}
loc2 <- locus(SLE_gwas_sub, gene = 'STAT4', flank = 1e5, LD = "r2")
loc3 <- locus(SLE_gwas_sub, gene = 'IRF5', flank = c(7e4, 2e5), LD = "r2")

# multiplot
pdf("myplot.pdf", height = 7, width = 10)
oldpar <- multi_layout(nrow = 1, ncol = 3)
plot(loc, use_layout = FALSE, legend_pos = 'topleft')
plot(loc2, use_layout = FALSE, legend_pos = NULL)
plot(loc3, use_layout = FALSE, legend_pos = NULL)
par(oldpar)  # reset to default
dev.off()
```

```{r, echo=FALSE, fig.dim = c(10,5)}
loc2 <- locus(SLE_gwas_sub, gene = 'STAT4', flank = 1e5, LD = "r2")
loc3 <- locus(SLE_gwas_sub, gene = 'IRF5', flank = c(7e4, 2e5), LD = "r2")

# multiplot
oldpar <- multi_layout(nrow = 1, ncol = 3)
plot(loc, use_layout = FALSE, legend_pos = 'topleft')
plot(loc2, use_layout = FALSE, legend_pos = NULL)
plot(loc3, use_layout = FALSE, legend_pos = NULL)
par(oldpar)  # reset to default
```

# Plot gene annotation only

The gene track can be plotted from a `locus` class object using the function 
`genetrack()`. This uses base graphics, so `layout()` can be used to stack 
custom-made plots above or below the gene tracks.

```{r}
genetracks(loc)
```

The function allows control over plotting of the gene tracks such as changing
the number of gene annotation tracks.

```{r}
# Limit the number of tracks
# Filter by gene biotype
# Customise colours
genetracks(loc, maxrows = 3, filter_gene_biotype = 'protein_coding',
           gene_col = 'grey', exon_col = 'orange', exon_border = 'darkgrey')
```

# Convert plots to grid objects

The `gridGraphics` package can be used to convert locus plots which are produced
in base graphics to `grid` objects (grobs), which can then be laid out on the
same page as `ggplot2` plots etc. The `gridBase` package allows `grid` and base
graphics outputs to be mixed.

