---
title: "locuszoomr"
author: "Myles Lewis"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
    toc_depth: 2
    number_sections: true
vignette: >
  %\VignetteIndexEntry{locuszoomr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE
)
library(locuszoomr)
```

# Introduction

The `locuszoomr` package allows users to produce publication ready gene locus 
plots very similar to those produced by the web interface locuszoom 
(http://locuszoom.org), but running purely locally in R. It provides 
customisation to the plots.

These gene annotation plots are produced via R base graphics system. A ggplot2
version is also available.

# Installation

Bioconductor packages `ensembldb` and an Ensembl database package are required 
before installation.

```{r eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("ensembldb")
BiocManager::install("EnsDb.Hsapiens.v75")
```

Install from Github
```{r eval = FALSE}
devtools::install_github("myles-lewis/locuszoomr")
```

`locuszoomr` automatically leverages the `LDlinkR` package to query 1000 Genomes 
for linkage disequilibrium (LD) across SNPs. In order to make use of this API
function you will need a personal access token (see the `LDlinkR` vignette), 
available from the LDlink website https://ldlink.nih.gov/?tab=apiaccess.

Requests to LDlink are cached using the `memoise` package, to reduce API 
requests. This is helpful when modifying plots for aesthetic reasons.

# Example locus plot

The quick example below uses a small subset (3 loci) of a GWAS dataset
incorporated into the package as a demo. The dataset is from a genetic study on
Systemic Lupus Erythematosus (SLE) by Bentham et al (2015). The full GWAS
summary statistics can be downloaded from
https://www.ebi.ac.uk/gwas/studies/GCST003156. The data format is shown below.

```{r}
library(locuszoomr)
data(SLE_gwas_sub)  ## limited subset of data from SLE GWAS
head(SLE_gwas_sub)
```

We plot a locus from this dataset by extracting a subset of the data using the
`locus()` function. Make sure you load the correct Ensembl database.

```{r warning=FALSE, message=FALSE, fig.dim=c(8, 7)}
library(EnsDb.Hsapiens.v75)
loc <- locus(SLE_gwas_sub, gene = 'UBE2L3', flank = 1e5)
summary(loc)
locus_plot(loc)
```

When `locus()` is called, the function tries to autodetect which columns in the
data object refer to chromosome, position, SNP/feature ID and p-value. These
columns can be specified manually using the arguments `chrom`, `pos`, `labs` and
`p` respectively.

# Controlling the locus

The genomic locus can be specified in several ways. The simplest is to specify a
gene by name/symbol using the `gene` argument. The location of the gene is
obtained from the specified Ensembl database. The amount of flanking regions can
either be controlled by specifying `flank` which defaults to 50kb either side of
the ends of the gene. `flank` can either be a single number or a vector of 2
numbers if different down/upstream flanking lengths are required. Alternatively
a fixed genomic window (eg. 1 Mb) centred on the gene of interest can be
specified using the argument `fix_window`. The locus can be specified manually
by specifying chromosome using `seqname` and genomic position range using
`xrange`.

# Obtaining LD information

Once an API personal access token has been obtained, the
[LDlink](https://ldlink.nih.gov/?tab=home) API can be called using the function
`link_LD()` to retrieve LD (linkage disequilibium) information at the locus
which is overlaid on the locus plot. This is shown as a colour overlay showing
the level of $r^2$ between SNPs and the index SNP which defaults to the SNP with
the lowest p-value (or the SNP can be specified manually).

```{r eval = FALSE}
# Locus plot using SLE GWAS data from Bentham et al 2015
# FTP download full summary statistics from
# https://www.ebi.ac.uk/gwas/studies/GCST003156
library(data.table)
SLE_gwas <- fread('../bentham_2015_26502338_sle_efo0002690_1_gwas.sumstats.tsv')
loc <- locus(SLE_gwas, gene = 'UBE2L3', flank = 1e5)
loc <- link_LD(loc, LDtoken = "your_token")
locus_plot(loc)
```

The subset of GWAS data included in the `locuszoomr` package has LD data already
acquired from LDlink which is included in the `r2` column. This can be plotted
by setting `LD = "r2"`. This method also allows users to add their own LD
information from their own datasets to loci.

```{r fig.dim=c(8, 7)}
loc <- locus(SLE_gwas_sub, gene = 'UBE2L3', flank = 1e5, LD = "r2")
locus_plot(loc)
```

# Plot customisation

Various plotting options can be customised through arguments via the call to
`locus_plot()`. Plot borders can be set using `border = TRUE`. The chromosome position
$x$ axis labels can be placed under the top or bottom plots using 
`xtick = "top"` or `"bottom"`.

See the help page at `?plot.locus` for more details.

# Customising gene tracks

The gene tracks can be also customised with colours and gene label text
position. See the help page at `?plot.locus` for more details.

```{r eval = FALSE}
# Filter by gene biotype
locus_plot(loc, filter_gene_biotype = "protein_coding")

# Custom selection of genes using gene names
locus_plot(loc, filter_gene_name = c('UBE2L3', 'RIMBP3C', 'YDJC', 'PPIL2',
                                     'PI4KAP2', 'MIR301B'))
```

# Plot gene annotation only

The gene track can be plotted from a `locus` class object using the function 
`genetrack()`. This uses base graphics, so `layout()` can be used to stack 
custom-made plots above or below the gene tracks.

```{r fig.dim=c(7, 3.5)}
genetracks(loc)
```

The function allows control over plotting of the gene tracks such as changing
the number of gene annotation tracks and the colour scheme. Set
`showExons=FALSE` to show only genes and hide the exons.

```{r fig.dim=c(7, 2.5)}
# Limit the number of tracks
# Filter by gene biotype
# Customise colours
genetracks(loc, maxrows = 3, filter_gene_biotype = 'protein_coding',
           gene_col = 'grey', exon_col = 'orange', exon_border = 'darkgrey')
```

# Add eQTL data via LDlink

GTEx eQTL data can be accessed via the LDlinkR package which queries the
[LDlink](https://ldlink.nih.gov/?tab=home) server API to download data and add
it to a 'locus' object. The eQTL data is stored in `LDexp` slot in the locus
object.

```{r eval=FALSE}
# obtain GTEx eQTL data from LDlinkR
# needs personal access token for LDlink
loc2 <- locus(SLE_gwas_sub, gene = 'IRF5', flank = c(7e4, 2e5), LD = "r2")
loc2 <- link_eqtl(loc2, LDtoken = "..")
head(loc2$LDexp)  # show eQTL data
```

Often at a locus eQTL data may be available on a range of genes and tissues. The
following code shows how to see which genes and tissues are available and how
eQTL SNPs they each have.

```{r eval=FALSE}
table(loc2$LDexp$Gene_Symbol)
##      AHCYL2           IRF5            KCP        METTL2B           NRF1   RP11-212P7.2 
##           5           1618              1              2              9              3 
## ...

table(loc2$LDexp$Tissue)
##             Adipose - Subcutaneous          Adipose - Visceral (Omentum) 
##                                 78                                    69 
##                      Adrenal Gland                        Artery - Aorta 
##                                 38                                   109
## ...
```

Embedded eQTL data can be plotted using `eqtl_plot()` or shown as an overlay
plot using `overlay_plot()` which shows GWAS or association data on the *y* axis
and overlays eQTL results using symbols and colours.

```{r eval=FALSE}
# just plot eQTL results
eqtl_plot(loc2, tissue = "Whole Blood", eqtl_gene = "IRF5")

# overlay eqtl plot
overlay_plot(loc2, eqtl_gene = "IRF5")
```

```{r, out.width='60%', fig.align="center", echo=FALSE}
knitr::include_graphics("eqtl_overlay.png")
```

# Change *y*-axis variable

Instead of plotting -log~10~ p-value on the y axis, it is possible to specify a
different variable in your dataset using the argument `yvar`.

```{r fig.dim=c(8, 7)}
locb <- locus(SLE_gwas_sub, gene = 'UBE2L3', flank = 1e5, yvar = "beta")
locus_plot(locb)
```

# Arrange multiple plots

`locuszoomr` uses `graphics::layout` to arrange plots. To layout multiple locus
plots side by side, use the function `multi_layout()` to set the number of locus
plots per row and column. Then set the argument `use_layout = FALSE` when
plotting locus objects. `multi_layout()` returns a list of original `par()`
parameters so they can be reset after plotting has finished.

```{r eval = FALSE}
loc2 <- locus(SLE_gwas_sub, gene = 'IRF5', flank = c(7e4, 2e5), LD = "r2")
loc3 <- locus(SLE_gwas_sub, gene = 'STAT4', flank = 1e5, LD = "r2")

# multiplot
pdf("myplot.pdf", width = 10, height = 7)
oldpar <- multi_layout(nrow = 1, ncol = 3)
locus_plot(loc, use_layout = FALSE, legend_pos = 'topleft')
locus_plot(loc2, use_layout = FALSE, legend_pos = NULL)
locus_plot(loc3, use_layout = FALSE, legend_pos = NULL)
par(oldpar)  # reset to default
dev.off()
```

```{r, echo=FALSE, fig.dim = c(10,5)}
loc2 <- locus(SLE_gwas_sub, gene = 'IRF5', flank = c(7e4, 2e5), LD = "r2")
loc3 <- locus(SLE_gwas_sub, gene = 'STAT4', flank = 1e5, LD = "r2")

# multiplot
oldpar <- multi_layout(nrow = 1, ncol = 3)
locus_plot(loc, use_layout = FALSE, legend_pos = 'topleft')
locus_plot(loc2, use_layout = FALSE, legend_pos = NULL)
locus_plot(loc3, use_layout = FALSE, legend_pos = NULL)
par(oldpar)  # reset to default
```

# Layering plots

## Column of plots

`locuszoomr` has been designed with modular functions to enable layering of
plots on top of each other in a column with gene tracks on the bottom.
`scatter_plot()` is used to generate the locus plot. `line_plot()` is used as an
example of an additional plot. Also see `eqtl_plot()` for plotting eQTL
information retrieved via LDlinkR.

```{r eval=FALSE}
pdf("myplot2.pdf", width = 6, height = 8)
# set up layered plot with 2 plots & a gene track; store old par() settings
oldpar <- set_layers(2)
scatter_plot(loc, xticks = FALSE)
line_plot(loc, col = "orange", xticks = FALSE)
genetracks(loc)
par(oldpar)  # revert par() settings
dev.off()
```

```{r, echo=FALSE, fig.dim = c(6,8)}
oldpar <- set_layers(2)
scatter_plot(loc, xticks = FALSE)
line_plot(loc, col = "orange", xticks = FALSE)
genetracks(loc)
par(oldpar)
```

## Overlaid plots

`scatter_plot()` can be called with argument `add = TRUE` to add multiple sets
of points overlaid on one plot.

```{r, echo=FALSE, message=FALSE, fig.dim = c(7,7)}
dat <- SLE_gwas_sub
dat$p2 <- -log10(dat$p * 0.1)
locp <- locus(dat, gene = 'UBE2L3', flank = 1e5)
locp2 <- locus(dat, gene = 'UBE2L3', flank = 1e5, yvar = "p2")

# set up overlaid plot with 1 plot & a gene track; store old par() settings
oldpar <- set_layers(1)
scatter_plot(locp, xticks = FALSE, pcutoff = NULL, ylim = c(0, 16))
scatter_plot(locp2, xticks = FALSE, pcutoff = NULL, chromCol = "orange", add = TRUE)
genetracks(loc)
par(oldpar)  # revert par() settings
```

# ggplot2 version

`ggplot2` versions of several of the above functions are available. For a whole
locus plot use `locus_ggplot()`.

```{r, eval=FALSE}
locus_ggplot(loc)
```

The gene tracks can be produced on their own as a grid package grob using
`gg_genetracks()`.

```{r, eval=FALSE}
grid::grid.newpage()
gg_genetracks(loc)
```

The scatter plot alone can be produced as a ggplot2 object.

```{r, eval=FALSE}
p <- gg_scatter(loc)
p
```

Finally, `gg_addgenes()` can be used to add gene tracks to an existing ggplot2
plot that has been previously created and customised.

```{r, eval=FALSE}
gg_addgenes(p, loc)
```

For users who prefer the grid system we recommend also looking at Bioconductor
packages `Gviz` or `ggbio` as possible alternatives.

# Arrange multiple ggplots

It is possible to use either the `cowplot` or `ggpubr` packages to layout
multiple locus ggplots on the same page.

```{r, fig.dim = c(12, 7)}
library(cowplot)
p1 <- locus_ggplot(loc, draw = FALSE)
p2 <- locus_ggplot(loc2, legend_pos = NULL, draw = FALSE)
plot_grid(p1, p2, ncol = 2)
```

Or using ggpubr or gridExtra:

```{r eval=FALSE}
library(ggpubr)
pdf("my_ggplot.pdf", width = 10)
ggarrange(p1, p2, ncol = 2)
dev.off()

library(gridExtra)
pdf("my_ggplot.pdf", width = 10)
grid.arrange(p1, p2, ncol = 2)
dev.off()
```

# References

Pruim RJ, Welch RP, Sanna S, Teslovich TM, Chines PS, Gliedt TP, Boehnke M,
Abecasis GR, Willer CJ. (2010) LocusZoom: Regional visualization of genome-wide
association scan results. *Bioinformatics* 2010; 26(18): 2336-7.
